[ZEPPELIN-2152] Added proxy config support in Helium bundle factory

necosta opened the pull request:
### What is this PR for?
Proxy configuration implementation in Helium bundle factory. Used by Node, NPM and YarnPkg installers.


### What type of PR is it?
[Improvement]

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-2152

### How should this be tested?
Trigger Helium plugins behind proxy

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

necosta commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/helium/HeliumBundleFactory.java:
```diff
@@ -129,6 +132,7 @@ void installNodeAndNpm() throws TaskRunnerException {
       YarnInstaller yarnInstaller = frontEndPluginFactory.getYarnInstaller(getProxyConfig());
       yarnInstaller.setYarnVersion(YARN_VERSION);
       yarnInstaller.install();
+      yarnCacheDir.mkdirs();
```
This allows me to have unit tests passing. Raised on https://issues.apache.org/jira/browse/ZEPPELIN-2469, pull request https://github.com/apache/zeppelin/pull/2362

1ambda commented:
Can we add documentation for this? under the security section or somewhere appropriate.

![image](https://cloud.githubusercontent.com/assets/4968473/26378753/ec02b244-4050-11e7-9701-ffa65ebb7aaf.png)

- To add a new section, Modify https://github.com/apache/zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L115-L118
- For content, write a md file into https://github.com/apache/zeppelin/tree/master/docs/security

1ambda commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/helium/HeliumBundleFactory.java:
```diff
@@ -140,8 +144,43 @@ void installNodeAndNpm() throws TaskRunnerException {
+    try {
+      // Order matters, first tries secure proxy
+      proxies.add(generateProxy("secure", new URI(httpsProxy)));
+      proxies.add(generateProxy("insecure",new URI(httpProxy)));
```
[ERROR] HeliumBundleFactory.java[158:44] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Zeppelin ........................................... SUCCESS [  3.897 s]
[INFO] Zeppelin: Interpreter .............................. SUCCESS [  9.132 s]
[INFO] Zeppelin: Zengine .................................. FAILURE [  1.260 s]

1ambda commented:
One more suggestion. Can we support proxy for helium registry file?

https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/helium/HeliumOnlineRegistry.java#L68

necosta commented:
@1ambda ,
* Added documentation although I'm unclear if that is what you are after. 
* Fixed bug where secure proxy was used for insecure connections (now http proxy for http connections, https proxy for https connections)
* Added proxy config for Helium Online registry

Leemoonsoo commented:
Thanks @necosta for great contribution!

LGTM and merge to master and branch-0.7 if no further discussions.

Leemoonsoo commented:
@necosta Please let me know your ASF JIRA id, to assign ZEPPELIN-2152 and close

necosta commented:
Thanks @Leemoonsoo for accepting the pull request! 
If I understood correctly, my ASF JIRA id/username is nelsonc.

1ambda commented:
Helium feature doesn't work after this PR.

- cloned new zeppelin repository
- built ` mvn clean package -DskipTests;`
- started zeppelin instance
- then got this error.

```
ERROR [2017-05-30 17:23:55,117] ({main} ZeppelinServer.java[<init>]:145) - 
java.lang.NullPointerException
	at java.net.URI$Parser.parse(URI.java:3023)
	at java.net.URI.<init>(URI.java:595)
	at org.apache.zeppelin.helium.HeliumOnlineRegistry.getProxy(HeliumOnlineRegistry.java:123)
	at org.apache.zeppelin.helium.HeliumOnlineRegistry.getAll(HeliumOnlineRegistry.java:71)
	at org.apache.zeppelin.helium.Helium.getAllPackageInfo(Helium.java:190)
```
